File: apps/android/src/main/kotlin/dev/supermux/android/editor/EditorState.kt
package dev.supermux.android.editor

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import dev.supermux.net.FsDiffResult
import dev.supermux.net.RepoDiff
import dev.supermux.net.ReviewComment
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

class EditorTab(path: String, content: String) {
    val path = path
    var content by mutableStateOf(content)
    var savedContent by mutableStateOf(content)
    var scrollTop by mutableStateOf(0)
}

class EditorState(
    private val fsRead: suspend (String) -> Result<String>,
    private val fsWrite: suspend (String, String) -> Boolean,
    private val scope: CoroutineScope,
) {
    val tabs = mutableStateListOf<EditorTab>()
    var activeTabPath by mutableStateOf<String?>(null)
    var loadingPath by mutableStateOf<String?>(null)
    var loadError by mutableStateOf<String?>(null)
    var saving by mutableStateOf(false)

    /** File tree UI state — survives panel / session switches while composed. */
    val treeRoot = mutableStateListOf<TreeNode>()
    var treeRootLoaded by mutableStateOf(false)
    var expandedPaths by mutableStateOf(setOf<String>())
    var treeLoadingPaths by mutableStateOf(setOf<String>())
    var treeVisible by mutableStateOf<Boolean?>(null)
    var searchQuery by mutableStateOf("")

    /** Markdown preview toggle (Eye/Pencil). Single flag like iOS — gated by activeIsMarkdown. */
    var previewMode by mutableStateOf(false)

    // Diff / inline code-review state — per-session, survives panel switches like the tree.
    var showDiff by mutableStateOf(false)
    var diffRepos by mutableStateOf<List<RepoDiff>>(emptyList())
    var diffComments by mutableStateOf<List<ReviewComment>>(emptyList())
    var diffLoading by mutableStateOf(false)

    /** Workdir-relative paths the broker reported changed on disk (fs_changed) → reload banner. */
    var changedPaths by mutableStateOf(setOf<String>())

    val activeTab: EditorTab?
        get() = tabs.find { it.path == activeTabPath }

    fun isDirty(path: String): Boolean {
        val tab = tabs.find { it.path == path } ?: return false
        return tab.content != tab.savedContent
    }

    fun captureActiveScroll(scrollTop: Int) {
        activeTab?.scrollTop = scrollTop
    }

    fun openFile(path: String) {
        tabs.find { it.path == path }?.let {
            activeTabPath = path
            loadError = null
            return
        }
        loadingPath = path
        loadError = null
        scope.launch {
            fsRead(path)
                .onSuccess { content ->
                    tabs.add(EditorTab(path, content))
                    activeTabPath = path
                    loadingPath = null
                }
                .onFailure { err ->
                    loadError = err.message ?: "Could not open file"
                    loadingPath = null
                }
        }
    }

    fun closeTab(path: String) {
        val idx = tabs.indexOfFirst { it.path == path }
        if (idx == -1) return
        tabs.removeAt(idx)
        if (activeTabPath == path) {
            activeTabPath = tabs.getOrNull(idx.coerceAtMost(tabs.lastIndex))?.path
        }
        if (activeTabPath == null) loadError = null
    }

    fun selectTab(path: String) {
        activeTabPath = path
        loadError = null
    }

    fun updateContent(path: String, content: String) {
        tabs.find { it.path == path }?.content = content
    }

    fun saveActive() {
        val tab = activeTab ?: return
        if (saving) return
        saving = true
        scope.launch {
            if (fsWrite(tab.path, tab.content)) {
                tab.savedContent = tab.content
            }
            saving = false
        }
    }

    // ── Diff / inline code-review (ports EditorState.swift:61-77) ───────────────

    /** Fetch the diff; only flip [showDiff] on a non-null result so a failed fetch never
     *  opens an empty pane (parity EditorState.swift:67). */
    suspend fun loadDiff(fsDiff: suspend () -> FsDiffResult?) {
        diffLoading = true
        val res = fsDiff()
        diffLoading = false
        if (res == null) return
        diffRepos = res.repos
        diffComments = res.comments
        showDiff = true
    }

    /** Re-fetch the diff in place (after add/resolve/submit) — does not toggle [showDiff]. */
    suspend fun reloadDiff(fsDiff: suspend () -> FsDiffResult?) {
        val res = fsDiff() ?: return
        diffRepos = res.repos
        diffComments = res.comments
    }

    // ── Live file-watch reload (ports EditorState.swift:79-84, 130-144) ─────────

    /** Record disk-change notifications (workdir-relative paths, leading slash optional). */
    fun markChanged(paths: List<String>) {
        changedPaths = changedPaths + paths.map(::normPath)
    }

    fun isStale(path: String): Boolean = normPath(path) in changedPaths

    private fun normPath(p: String): String = p.removePrefix("/")

    /** Re-read a tab from disk and clear its stale flag (parity EditorState.swift:130-144). */
    suspend fun reload(path: String, fsRead: suspend (String) -> Result<String>) {
        val tab = tabs.find { it.path == path } ?: return
        loadingPath = path
        fsRead(path)
            .onSuccess { content ->
                tab.content = content
                tab.savedContent = content
                changedPaths = changedPaths - normPath(path)
            }
            .onFailure { err -> loadError = err.message ?: "Could not reload file" }
        loadingPath = null
    }
}

File: apps/iosApp/Supermux/Editor/EditorState.swift
import Foundation
import Shared

/// Source of truth for the editor pane — open tabs, their content/dirty state, and
/// the file-tree UI state. Mirrors the Android `EditorState`, but with UNLIMITED tabs
/// (no cap, no eviction — overrides the PWA's MAX_TABS=10). The WKWebView is a pushed
/// renderer; this type owns the data. `fsRead`/`fsWrite` are injected so it stays
/// testable and decoupled from `BrokerSession`.
@MainActor
@Observable
final class EditorState {
    struct Tab: Identifiable {
        let path: String
        var content: String
        var savedContent: String
        var scrollTop: Int
        var id: String { path }
        var isDirty: Bool { content != savedContent }
    }

    /// A pending request to scroll the editor to a line once a tab is active (chat tap →
    /// open at a line). The nonce makes a repeat of the same target a distinct value so the
    /// webview re-applies it; mirrors Android's `EditorTab.revealLine`.
    struct RevealRequest: Equatable {
        let path: String
        let line: Int
        let endLine: Int?
        let nonce: Int
    }

    private(set) var tabs: [Tab] = []
    var activeTabPath: String?
    private(set) var loadingPath: String?
    private(set) var loadError: String?
    private(set) var saving = false

    // Tree UI state — persists across pane / session switches.
    var treeRoot: [TreeNode] = []
    var treeRootLoaded = false
    var expandedPaths: Set<String> = []
    var treeLoadingPaths: Set<String> = []
    var treeVisible = true
    var searchQuery = ""

    // Diff / code-review state — per-session, survives switches like the tree.
    var showDiff = false
    private(set) var diffRepos: [RepoDiff] = []
    private(set) var diffComments: [ReviewComment] = []
    private(set) var diffLoading = false

    // Paths the broker reported changed on disk (fs_changed) → drives a reload banner.
    private(set) var changedPaths: Set<String> = []

    // Pending reveal-line request (chat tap → open at a line). Read by EditorPane to drive
    // the webview's cmRevealLine; `revealNonce` keeps repeat taps distinct.
    private(set) var reveal: RevealRequest?
    private var revealNonce = 0

    let sessionId: String
    private let fsRead: (String) async throws -> String
    private let fsWrite: (String, String) async -> Bool
    private let fsDiff: (() async -> FsDiffResult?)?

    var activeTab: Tab? { tabs.first { $0.path == activeTabPath } }

    init(sessionId: String,
         fsRead: @escaping (String) async throws -> String,
         fsWrite: @escaping (String, String) async -> Bool,
         fsDiff: (() async -> FsDiffResult?)? = nil) {
        self.sessionId = sessionId
        self.fsRead = fsRead
        self.fsWrite = fsWrite
        self.fsDiff = fsDiff
    }

    func loadDiff() async {
        guard let fsDiff else { return }
        diffLoading = true
        let res = await fsDiff()
        diffLoading = false
        // Only open the diff pane on success — never show an empty/stale diff for a failed fetch.
        guard let res else { return }
        diffRepos = res.repos
        diffComments = res.comments
        showDiff = true
    }

    func reloadDiff() async {
        guard let fsDiff, let res = await fsDiff() else { return }
        diffRepos = res.repos
        diffComments = res.comments
    }

    /// Record disk-change notifications (workdir-relative paths, leading slash optional).
    func markChanged(_ paths: [String]) {
        for p in paths { changedPaths.insert(Self.normPath(p)) }
    }
    func isStale(_ path: String) -> Bool { changedPaths.contains(Self.normPath(path)) }
    private static func normPath(_ p: String) -> String { p.hasPrefix("/") ? String(p.dropFirst()) : p }

    func openFile(_ path: String) async {
        if tabs.contains(where: { $0.path == path }) {
            activeTabPath = path
            loadError = nil
            return
        }
        loadingPath = path
        loadError = nil
        do {
            let content = try await fsRead(path)
            tabs.append(Tab(path: path, content: content, savedContent: content, scrollTop: 0))
            activeTabPath = path
        } catch {
            loadError = error.localizedDescription
        }
        loadingPath = nil
    }

    /// Open [path] then, once the tab is active, request a scroll to [line] (1-indexed).
    /// A nil [line] just opens the file. Mirrors Android's `EditorState.openFileAtLine`.
    func openFileAtLine(_ path: String, line: Int?, endLine: Int?) {
        Task {
            await openFile(path)
            // Reveal only if the open produced a tab (a failed fsRead adds none) and a line was given —
            // avoids a stale-nonce reveal firing later when the path is reopened from the tree.
            guard tabs.contains(where: { $0.path == path }), let line else { return }
            revealNonce += 1
            reveal = RevealRequest(path: path, line: line, endLine: endLine, nonce: revealNonce)
        }
    }

    func closeTab(_ path: String) {
        guard let idx = tabs.firstIndex(where: { $0.path == path }) else { return }
        tabs.remove(at: idx)
        if activeTabPath == path {
            activeTabPath = tabs.isEmpty ? nil : tabs[min(idx, tabs.count - 1)].path
        }
        if activeTabPath == nil { loadError = nil }
    }

    func updateContent(_ path: String, _ s: String) {
        guard let idx = tabs.firstIndex(where: { $0.path == path }) else { return }
        tabs[idx].content = s
    }

    func saveActive() async {
        guard !saving, let path = activeTabPath,
              let idx = tabs.firstIndex(where: { $0.path == path }) else { return }
        saving = true
        let content = tabs[idx].content
        if await fsWrite(path, content),
           let i = tabs.firstIndex(where: { $0.path == path }) {
            tabs[i].savedContent = content
        }
        saving = false
    }

    func reload(_ path: String) async {
        guard tabs.contains(where: { $0.path == path }) else { return }
        loadingPath = path
        do {
            let content = try await fsRead(path)
            if let idx = tabs.firstIndex(where: { $0.path == path }) {
                tabs[idx].content = content
                tabs[idx].savedContent = content
            }
            changedPaths.remove(Self.normPath(path))
        } catch {
            loadError = error.localizedDescription
        }
        loadingPath = nil
    }

    func setScroll(_ path: String, _ top: Int) {
        guard let idx = tabs.firstIndex(where: { $0.path == path }) else { return }
        tabs[idx].scrollTop = top
    }
}

struct TreeNode: Identifiable {
    let entry: FsEntry
    let path: String
    var children: [TreeNode]?
    var loaded: Bool
    var id: String { path }
    var isDir: Bool { children != nil }
}
